### PR TITLE
IGNITE-21178 Add delay duration when waiting for indexes

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-21178

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java [hunk 1]
```diff
@@ -315,28 +319,50 @@ public void checkExplicitTxRollback() {
...
+            @Values(booleans = {true, false}) boolean roTx,
+            @Values(booleans = {true, false}) boolean commit,
+            @Values(booleans = {true, false}) boolean explicit
+    ) throws InterruptedException {
         sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
         sql("CREATE INDEX TEST_IDX ON TEST(VAL0)");
 
+        // Wait for the index to be created.
+        assertTrue(waitForCondition(
+                () -> CLUSTER.runningNodes().allMatch(node -> {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:348]: Why you should waiting appear index for all nodes, as I understand it should be enough wait just publish state for one node.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java [hunk 2]
```diff
@@ -315,28 +319,50 @@ public void checkExplicitTxRollback() {
...
+    public void checkMixedTransactionsForIndex(
+            @Values(booleans = {true, false}) boolean roTx,
+            @Values(booleans = {true, false}) boolean commit,
+            @Values(booleans = {true, false}) boolean explicit
+    ) throws InterruptedException {
         sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
         sql("CREATE INDEX TEST_IDX ON TEST(VAL0)");
 
+        // Wait for the index to be created.
+        assertTrue(waitForCondition(
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:347]: please check `org.apache.ignite.internal.ClusterPerClassIntegrationTest#sql(int, org.apache.ignite.tx.Transaction, java.lang.String, java.lang.Object[])`. It already should awaiting index availability

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java [hunk 3]
```diff
@@ -315,28 +319,50 @@ public void checkExplicitTxRollback() {
...
 
-        checkMixedTransactions(planMatcher);
+        checkMixedTransactions(planMatcher, roTx, commit, explicit);
     }
 
 
     /** Check correctness of rw and ro transactions for index scan. */
-    @Test
-    public void checkMixedTransactionsForIndex() throws Exception {
+    @CartesianTest(name = "roTx={0} commit={1} explicit={2}")
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:337]: Yes, but I think it's easier to understand what happens if a test fails (all these combinations describe different tests states and IMHO should be separated). I can revert this change, if you insist.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java (hunk 1 above)
Comment (sashapolo) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:348]: Can you explain why?

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java (hunk 2 above)
Comment (sashapolo) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:347]: Yes, but in this test we call `assertQuery` to check the plan, which does not use the `sql` method

Comment (ygerzhedovich) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:347]: We awaiting index availability after execute create an index. So `assertQuery` doesn't require additional awaiting.

Comment (sashapolo) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:347]: Oh, I understand, what you are talking about. Let me investigate, why that check doesn't work

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java (hunk 1 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java:348]: publication in the catalog should occur simultaneously for all nodes

Comment (sashapolo) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java]: I don't think that's true but that's not relevant anymore

**Review by lowka — APPROVED** [review]

**Pull request merged**